### PR TITLE
Fix merging browser keys

### LINKED ISSUE #4077: Merge from database overwrites Browser Integration Stored Keys in Database settings
When merging from another database that has one stored key/value pair to the existing database that has a key/value pair with a different key, the end result is that the merged database has the single key/value pair from the "from" database.

## Expected Behavior
I expect the set of keys in the merged database to be the union of the sets of keys from the from and to databases.  When the from and to databases both have the same key but with different values, either the user should be prompted to choose which value or the value changed more recently should be the value in the merged database

## Current Behavior
The set of keys in the merged database is equal to the set in the from database.

## Possible Solution
described in expected behavior

## Steps to Reproduce
[NOTE]: # ( Provide a link to a live example, or an unambiguous set of steps to )
[NOTE]: # ( reproduce this bug. Include code to reproduce, if relevant )
1. start with identical databases on two different machines where the database settings under Browser Integration has no stored keys
2. connect to keepassxc using the keepassxc-browser plugin on machine A.  when prompted to give a name for the connection, use the value "A"
3. save the database on machine A
4. connect to keepasxc using the keepassxc-browser plugin on machine B.  when prompted to give a name for the connection, use the value "B"
5. save the database on machine B
6. transfer a copy of the database from machine A to machine B
7. with the machine B database opened in keepassxc on machine B, select Database->Merge from database...  In the file chooser, selec the copy of the database from machine A described in step 6
8. observe that the resulting merged database has only key A in Database->Database settings under Browser Integration.  key B is missing

## Context
I use git to manage copies of my database on different machines.  When I make changes to a database on one machine (and not on the other), I update my copy  on the other machine by using git to pull the changes.  If I make changes on two different machines (now much more common with keepassxc-browser saving settings in the database related to which entries to use for which URLs), I make commits on both machines.  To merge tha, I do a git merge --no-checkin -s ours.  But before I check in the result of the git merge, I perform a "Database->Merge from database...." and specify a copy of the remote machine's copy (i.e. the "theirs" copy) of the database.  This allows me to actually merge the two copies and to also correctly reflect that merge in git.

## Debug Info
[NOTE]: # ( Paste debug info from Help → About here )
KeePassXC - 2.5.1
Revision: 0fd8836

Libraries:
- Qt 5.12.3

Operating system: Gentoo Linux and Windows 10 (from database of the merge is Gentoo Linux)
CPU architecture: x86_64 on both
Kernel: linux 5.1.15

Enabled extensions: (in French because that's my system language)
- Saisie automatique (this is auto-type)
- Intégration aux navigateurs (this is browser integration)
- Agent SSH
- YubiKey
- Intégration au Secret Service

### LINKED ISSUE #4684: KeeShare import causes Browser Integration entries to change or disappear
## Overview

I use KeeShare. I have a main database D1 with Browser Integration key A. I have a KeeShare "Import from" entry for database D2, with browser integration key B.

After an update and save of database D2, the browser integration keys in D1 are now just left with B. A has been completely deleted.

If database D2 has no browser integration keys, then upon a save of D2, the keys from database D1 are also deleted.

## Steps to Reproduce
1. Create two databases and use Keeshare "import only" mode to import the second database into the first.
2. Create Keeshare entries in the first database.
3. Modify the second database and save it.

KeePassXC - Version 2.5.4
Revision: dcca5aa

Operating System: Linux
Desktop Env: KDE
Windowing System: X11

### PR DESCRIPTION
**Pull request by varjolintu**
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Explain large or complex code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" )
When importing data from another database via merge, do not delete any old keys from the destination.

Fixes #4077.
Fixes #4684.

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Bug fix (non-breaking change that fixes an issue)

### CODE REVIEW

**Comment by varjolintu** [comment]
As the Ubuntu build fails, I was thinking if it would be better to move the browser key prefix from BrowserService to CustomData class instead, as `CustomData::BrowserKeyPrefix`. In that way we don't have to include any of the browser files to Merger.

File: src/core/Merger.cpp [hunk 1]
```diff
@@ -632,7 +633,9 @@ Merger::ChangeList Merger::mergeMetadata(const MergeContext& context)
 
         // Check missing keys from source. Remove those from target
         for (const auto& key : targetCustomDataKeys) {
-            if (!sourceMetadata->customData()->contains(key)) {
+            // Do not remove any browser keys
+            if (!sourceMetadata->customData()->contains(key) && !key.startsWith(BrowserService::ASSOCIATE_KEY_PREFIX)
+                && !key.startsWith(CustomData::Created)) {
```
Comment (droidmonkey) [inline src/core/Merger.cpp:638]: I don't like how specific this is to the browser plugin. I would prefer a new function in the Merger class, perhaps named `isProtectedCustomData(key)`, that does these checks and is separate from the mechanics of the merge.

What is the long-term solution to this as more and more plugins use the custom data?

Comment (varjolintu) [reply src/core/Merger.cpp:638]: That would be much nicer, yes.

For the long-term solution I would suggest creating a new class for a certain type of custom data that will have Created/Modified timestamps (and a possible prefix), so merging identical key/values could be compared, and the implementation would be slightly more generic.

**Comment by droidmonkey** [comment]
I love this new solution

**Comment by varjolintu** [comment]
> I love this new solution, we can use the new BrowserShared namespace once the browser refactor is merged.

That was exactly my idea as well.

**Pull request merged**